docs/src/tutorials/nonlinear/operator_ad.jl

```diff
@@ -0,0 +1,314 @@
+    for i in 1:N, j in 1:i
+        H[i, j] = h[i, j]
+    end
+    return
+end
+
+# Let's check that we find the analytic solution:
+
+fdiff_H = zeros(2, 2)
+fdiff_∇²f(fdiff_H, x...)
+Test.@test ≈(analytic_H, fdiff_H)
+
+# ### JuMP example
+
+# The code for computing the gradient and Hessian using ForwardDiff can be
+# re-used for many operators. Thus, it is helpful to encapsulate it into the
+# function:
+
+"""
+    fdiff_derivatives(f::Function) -> Tuple{Function,Function}
+
+Return a tuple of functions that evalute the gradient and Hessian of `f` using
+ForwardDiff.jl.
+"""
+function fdiff_derivatives(f::Function)
+    function ∇f(g::AbstractVector{T}, x::Vararg{T,N}) where {T,N}
+        ForwardDiff.gradient!(g, y -> f(y...), collect(x))
+        return
+    end
+    function ∇²f(H::AbstractMatrix{T}, x::Vararg{T,N}) where {T,N}
+        h = ForwardDiff.hessian(y -> f(y...), collect(x))
+        for i in 1:N, j in 1:i
+            H[i, j] = h[i, j]
+        end
+        return
+    end
+    return ∇f, ∇²f
+end
+
+# Here's an example using `fdiff_derivatives`:
+
+function fdiff_rosenbrock()
+    model = Model(Ipopt.Optimizer)
+    set_silent(model)
+    @variable(model, x[1:2])
+    @operator(model, op_rosenbrock, 2, f, fdiff_derivatives(f)...)
+    @objective(model, Min, op_rosenbrock(x[1], x[2]))
+    optimize!(model)
+    Test.@test is_solved_and_feasible(model)
+    return value.(x)
+end
+
+fdiff_rosenbrock()
+
+# ## Enzyme
+
+# Another library for automatic differentiation in Julia is [Enzyme.jl](https://github.com/EnzymeAD/Enzyme.jl).
+
+# ### Pros and cons
+
+# The main benefit of Enzyme is that it can produce fast derivatives for
+# functions with many input arguments.
+
+# The main downsides are that it may throw unusual errors if your code uses an
+# unsupported feature of Julia and that it may have large compile times.
+
+# !!! warning
+#     The JuMP developers cannot help you debug error messages related to
+#     Enzyme. If the operator works, it works. If not, we suggest you try a
+#     different automatic differentiation library.
+
+# ### Gradient
+
+# The gradient can be computed using `ForwardDiff.autodiff` with the
+# `Enzyme.Reverse` mode. We need to wrap `x` in `Enzyme.Active` to indicate that
+# we want to compute the derivatives with respect to these arguments.
+
+function enzyme_∇f(g::AbstractVector{T}, x::Vararg{T,N}) where {T,N}
+    g .= Enzyme.autodiff(Enzyme.Reverse, f, Enzyme.Active.(x)...)[1]
+    return
+end
+
+# Let's check that we find the analytic solution:
+
+enzyme_g = zeros(2)
+enzyme_∇f(enzyme_g, x...)
+Test.@test ≈(analytic_g, enzyme_g)
+
+# ### Hessian
+
+# We can compute the Hessian in Enzyme using forward-over-reverse automatic
+# differentiation.
+
+# The code to implement the Hessian in Enzyme is complicated, so we will not
+# explain it in detail; see the [Enzyme documentation](https://enzymead.github.io/Enzyme.jl/v0.11.19/generated/autodiff/#Vector-forward-over-reverse).
```

Comment:
Outdated link: https://enzymead.github.io/Enzyme.jl/stable/

Reply:
I'd prefer to use specific links so that we don't risk links breaking with an unrelated Enzyme release.

Reply:
fair